File: apps/api/internal/service/ai.go
package service

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type AIGateway struct {
	BaseURL string
	Client  *http.Client
}

func NewAIGateway(baseURL string) AIGateway {
	return AIGateway{
		BaseURL: strings.TrimRight(baseURL, "/"),
		Client: &http.Client{
			Timeout: 3 * time.Second,
		},
	}
}

func (g AIGateway) EngineName() string {
	if g.BaseURL != "" {
		return "contract_ai_service"
	}
	return "local_go_fallback_ai"
}

func (g AIGateway) RemoteEnabled() bool {
	return g.BaseURL != ""
}

func (g AIGateway) ProcessSignals(signals []map[string]any) map[string]any {
	if g.BaseURL == "" {
		return fallbackProcessSignals(signals)
	}

	regionID := regionFromSignals(signals)
	risk, err := g.postJSON("/infer/risk", map[string]any{
		"region_id": regionID,
		"at_time":   time.Now().UTC().Format(time.RFC3339),
		"signals":   signals,
	})
	if err != nil {
		return fallbackProcessSignals(signals)
	}

	explain, err := g.postJSON("/infer/explain", map[string]any{
		"region_id":   regionID,
		"risk_score":  number(risk["risk_score"]),
	})
	if err != nil {
		return fallbackProcessSignals(signals)
	}

	interventions, err := g.postJSON("/infer/interventions", map[string]any{
		"region_id":  regionID,
		"risk_score": number(risk["risk_score"]),
	})
	if err != nil {
		return fallbackProcessSignals(signals)
	}

	drivers := sliceOfMaps(risk["top_drivers"])
	items := sliceOfMaps(interventions["interventions"])

	indicators := make([]map[string]any, 0, len(drivers))
	for _, driver := range drivers {
		name := stringValue(driver["name"])
		indicators = append(indicators, map[string]any{
			"type":        strings.ToLower(strings.ReplaceAll(name, " ", "_")),
			"severity":    round(number(driver["contribution"])),
			"description": fmt.Sprintf("%s driving risk in %s; direction is %s", name, regionID, stringValue(driver["direction"])),
		})
	}

	recommendations := make([]map[string]any, 0, len(items))
	for index, item := range items {
		recommendations = append(recommendations, map[string]any{
			"action":           stringValue(item["category"]),
			"priority":         index + 1,
			"estimated_impact": stringValue(item["expected_impact"]),
		})
	}

	return map[string]any{
		"assessment": map[string]any{
			"threat_level": int(round(number(risk["risk_score"]) * 100)),
			"confidence":   number(risk["confidence"]),
			"risk_factors": driverNames(drivers),
		},
		"indicators":      indicators,
		"recommendations": recommendations,
		"metadata": map[string]any{
			"model_version":             stringValue(risk["model_version"]),
			"processing_time":           0.12,
			"engine":                    "contract_ai_service",
			"feature_snapshot_timestamp": risk["feature_snapshot_timestamp"],
			"known_data_gaps":           risk["known_data_gaps"],
			"explanation_summary":       stringValue(explain["summary"]),
		},
	}
}

func (g AIGateway) PredictRisk(region, timeframe string) map[string]any {
	if g.BaseURL == "" {
		return fallbackPredict(region, timeframe)
	}

	risk, err := g.postJSON("/infer/risk", map[string]any{
		"region_id": region,
		"at_time":   time.Now().UTC().Format(time.RFC3339),
		"signals":   []map[string]any{},
	})
	if err != nil {
		return fallbackPredict(region, timeframe)
	}

	days := parseDays(timeframe)
	base := int(round(number(risk["risk_score"]) * 100))
	predictions := make([]map[string]any, 0, min(days, 7))
	for offset := 0; offset < min(days, 7); offset++ {
		predictions = append(predictions, map[string]any{
			"date":       time.Now().UTC().AddDate(0, 0, offset).Format("2006-01-02"),
			"risk_level": min(base+offset, 99),
			"confidence": maxFloat(0.5, round(number(risk["confidence"])-(float64(offset)*0.02))),
		})
	}

	return map[string]any{
		"region":      region,
		"timeframe":   timeframe,
		"predictions": predictions,
		"confidence":  number(risk["confidence"]),
		"factors":     driverNames(sliceOfMaps(risk["top_drivers"])),
		"metadata": map[string]any{
			"model_version": stringValue(risk["model_version"]),
			"engine":        "contract_ai_service",
		},
	}
}

func (g AIGateway) AnalyzeDrivers(payload map[string]any) map[string]any {
	if g.BaseURL == "" {
		return fallbackAnalyzeDrivers(payload)
	}

	explain, err := g.postJSON("/infer/explain", map[string]any{
		"region_id":  stringOrDefault(payload["region_id"], "national"),
		"risk_score": numberOrDefault(payload["risk_score"], 0.6),
	})
	if err != nil {
		return fallbackAnalyzeDrivers(payload)
	}

	driversRaw := sliceOfMaps(explain["top_drivers"])
	drivers := make([]map[string]any, 0, len(driversRaw))
	for _, item := range driversRaw {
		trend := "stable"
		if stringValue(item["direction"]) == "up" {
			trend = "increasing"
		}
		drivers = append(drivers, map[string]any{
			"name":         stringValue(item["name"]),
			"contribution": round(number(item["contribution"])),
			"trend":        trend,
			"confidence":   0.84,
		})
	}

	relationships := []map[string]any{}
	if len(drivers) >= 2 {
		relationships = append(relationships, map[string]any{
			"from":     drivers[0]["name"],
			"to":       drivers[1]["name"],
			"strength": 0.61,
		})
	}

	return map[string]any{
		"drivers":               drivers,
		"causal_relationships":  relationships,
		"metadata": map[string]any{
			"model_version":    stringValue(explain["model_version"]),
			"engine":           "contract_ai_service",
			"uncertainty_notes": explain["uncertainty_notes"],
		},
	}
}

func (g AIGateway) RecommendInterventions(region string, riskProfile map[string]any) map[string]any {
	if g.BaseURL == "" {
		return fallbackRecommend(region, riskProfile)
	}

	riskScore := numberOrDefault(riskProfile["risk_score"], numberOrDefault(riskProfile["threat_level"], 55)/100)
	response, err := g.postJSON("/infer/interventions", map[string]any{
		"region_id":  region,
		"risk_score": riskScore,
	})
	if err != nil {
		return fallbackRecommend(region, riskProfile)
	}

	items := sliceOfMaps(response["interventions"])
	interventions := make([]map[string]any, 0, len(items))
	for _, item := range items {
		interventions = append(interventions, map[string]any{
			"category":            stringValue(item["category"]),
			"expectedImpact":      stringValue(item["expected_impact"]),
			"timeToEffect":        stringValue(item["time_to_effect"]),
			"costBand":            "Policy-defined",
			"confidence":          stringValue(item["confidence"]),
			"effectiveness_score": 0.78,
			"constraintsApplied":  item["constraints_applied"],
		})
	}

	return map[string]any{
		"region":        region,
		"interventions": interventions,
		"metadata": map[string]any{
			"model_version": stringValue(response["model_version"]),
			"engine":        "contract_ai_service",
		},
	}
}

func (g AIGateway) postJSON(path string, payload map[string]any) (map[string]any, error) {
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest(http.MethodPost, g.BaseURL+path, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := g.Client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("upstream status %s", resp.Status)
	}

	var decoded map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&decoded); err != nil {
		return nil, err
	}
	return decoded, nil
}

func fallbackProcessSignals(signals []map[string]any) map[string]any {
	indicators := make([]map[string]any, 0, len(signals))
	for _, signal := range signals {
		signalType := stringOrDefault(signal["type"], "general")
		severity := scoreSignal(signalType, signal["data"])
		dataMap, _ := signal["data"].(map[string]any)
		keySummary := "limited telemetry"
		if len(dataMap) > 0 {
			keys := make([]string, 0, 2)
			for key := range dataMap {
				keys = append(keys, key)
				if len(keys) == 2 {
					break
				}
			}
			keySummary = strings.Join(keys, ", ")
		}
		indicators = append(indicators, map[string]any{
			"type":        signalType,
			"severity":    severity,
			"description": fmt.Sprintf("%s signal flagged from %s; severity scored at %.2f", strings.Title(strings.ReplaceAll(signalType, "_", " ")), keySummary, severity),
		})
	}

	threatLevel := 42
	if len(indicators) > 0 {
		total := 0.0
		for _, indicator := range indicators {
			total += number(indicator["severity"])
		}
		threatLevel = int(round((total / float64(len(indicators))) * 100))
	}

	riskFactors := make([]string, 0, min(len(indicators), 3))
	for index, indicator := range indicators {
		if index == 3 {
			break
		}
		riskFactors = append(riskFactors, strings.Title(strings.ReplaceAll(stringValue(indicator["type"]), "_", " ")))
	}
	if len(riskFactors) == 0 {
		riskFactors = []string{"Baseline monitoring"}
	}

	recommendations := []map[string]any{
		{"action": "Increase analyst review cadence", "priority": 1, "estimated_impact": impactBand(threatLevel)},
	}
	for _, indicator := range indicators {
		switch stringValue(indicator["type"]) {
		case "economic":
			recommendations = append(recommendations, map[string]any{"action": "Escalate economic stress monitoring", "priority": 2, "estimated_impact": "High"})
		case "environmental":
			recommendations = append(recommendations, map[string]any{"action": "Prepare resilience response coordination", "priority": 3, "estimated_impact": "Moderate"})
		}
	}

	return map[string]any{
		"assessment": map[string]any{
			"threat_level": threatLevel,
			"confidence":   round(minFloat(0.96, 0.58+(float64(len(indicators))*0.07))),
			"risk_factors": riskFactors,
		},
		"indicators":      indicators,
		"recommendations": recommendations,
		"metadata": map[string]any{
			"model_version": "local-dev-1.0",
			"processing_time": round(0.04 + (float64(len(signals)) * 0.015)),
			"engine": "local_go_fallback_ai",
			"timestamp": time.Now().UTC().Format(time.RFC3339),
		},
	}
}

func fallbackPredict(region, timeframe string) map[string]any {
	days := parseDays(timeframe)
	baseRisk := 48 + min(days, 14)
	predictions := make([]map[string]any, 0, min(days, 7))
	for offset := 0; offset < min(days, 7); offset++ {
		predictions = append(predictions, map[string]any{
			"date":       time.Now().UTC().AddDate(0, 0, offset).Format("2006-01-02"),
			"risk_level": min(95, baseRisk+offset),
			"confidence": maxFloat(0.62, round(0.85-(float64(offset)*0.02))),
		})
	}

	return map[string]any{
		"region":      region,
		"timeframe":   timeframe,
		"predictions": predictions,
		"confidence":  0.81,
		"factors":     []string{"Signal velocity", "Recent anomaly density", "Historical trend carry-over"},
		"metadata": map[string]any{
			"model_version": "local-dev-1.0",
			"engine":        "local_go_fallback_ai",
		},
	}
}

func fallbackAnalyzeDrivers(payload map[string]any) map[string]any {
	drivers := []map[string]any{
		{"name": "Baseline Stress", "contribution": 1.0, "trend": "increasing", "confidence": 0.84},
	}
	bestKey := ""
	bestValue := 0.0
	secondKey := ""
	secondValue := 0.0
	for key, raw := range payload {
		value := number(raw)
		if value <= 0 {
			continue
		}
		if value > bestValue {
			secondKey, secondValue = bestKey, bestValue
			bestKey, bestValue = key, value
		} else if value > secondValue {
			secondKey, secondValue = key, value
		}
	}

	if bestKey != "" {
		total := bestValue + secondValue
		if total == 0 {
			total = bestValue
		}
		drivers = []map[string]any{
			{"name": strings.Title(strings.ReplaceAll(bestKey, "_", " ")), "contribution": round(bestValue / total), "trend": "increasing", "confidence": 0.84},
		}
		if secondKey != "" {
			drivers = append(drivers, map[string]any{"name": strings.Title(strings.ReplaceAll(secondKey, "_", " ")), "contribution": round(secondValue / total), "trend": "stable", "confidence": 0.84})
		}
	}

	relationships := []map[string]any{}
	if len(drivers) >= 2 {
		relationships = append(relationships, map[string]any{
			"from":     drivers[0]["name"],
			"to":       drivers[1]["name"],
			"strength": 0.63,
		})
	}

	return map[string]any{
		"drivers":              drivers,
		"causal_relationships": relationships,
		"metadata": map[string]any{
			"model_version": "local-dev-1.0",
			"engine":        "local_go_fallback_ai",
		},
	}
}

func fallbackRecommend(region string, riskProfile map[string]any) map[string]any {
	threatLevel := int(numberOrDefault(riskProfile["threat_level"], 55))
	expectedImpact := "Moderate"
	if threatLevel >= 70 {
		expectedImpact = "High"
	}

	return map[string]any{
		"region": region,
		"interventions": []map[string]any{
			{
				"category":            "Targeted field assessment",
				"expectedImpact":      expectedImpact,
				"timeToEffect":        "Short",
				"costBand":            "Low",
				"confidence":          "High",
				"effectiveness_score": 0.83,
			},
			{
				"category":            "Multi-agency coordination cell",
				"expectedImpact":      "High",
				"timeToEffect":        "Medium",
				"costBand":            "Medium",
				"confidence":          "High",
				"effectiveness_score": 0.79,
			},
		},
		"metadata": map[string]any{
			"model_version": "local-dev-1.0",
			"engine":        "local_go_fallback_ai",
		},
	}
}

func regionFromSignals(signals []map[string]any) string {
	for _, signal := range signals {
		location, ok := signal["location"].(map[string]any)
		if !ok {
			continue
		}
		if region := stringValue(location["region"]); region != "" {
			return region
		}
		if name := stringValue(location["name"]); name != "" {
			return name
		}
	}
	return "national"
}

func driverNames(drivers []map[string]any) []string {
	names := make([]string, 0, len(drivers))
	for _, driver := range drivers {
		names = append(names, stringValue(driver["name"]))
	}
	return names
}

func parseDays(timeframe string) int {
	cleaned := strings.TrimSpace(strings.ToLower(timeframe))
	if strings.HasSuffix(cleaned, "d") {
		if value, err := strconv.Atoi(strings.TrimSuffix(cleaned, "d")); err == nil && value > 0 {
			return value
		}
	}
	return 7
}

func scoreSignal(signalType string, data any) float64 {
	score := 0.35
	dataMap, _ := data.(map[string]any)
	numericCount := 0
	numericTotal := 0.0
	textParts := make([]string, 0, len(dataMap))
	for _, raw := range dataMap {
		switch value := raw.(type) {
		case float64:
			numericTotal += value
			numericCount++
		case int:
			numericTotal += float64(value)
			numericCount++
		case string:
			textParts = append(textParts, strings.ToLower(value))
		}
	}
	if numericCount > 0 {
		score += minFloat((numericTotal/float64(numericCount))/200.0, 0.35)
	}
	textBlob := strings.Join(textParts, " ")
	for _, keyword := range []struct {
		word   string
		weight float64
	}{
		{"protest", 0.12},
		{"shortage", 0.10},
		{"violence", 0.18},
		{"drought", 0.09},
		{"price", 0.07},
		{"unemployment", 0.08},
	} {
		if strings.Contains(textBlob, keyword.word) {
			score += keyword.weight
		}
	}
	switch signalType {
	case "economic":
		score += 0.08
	case "environmental":
		score += 0.06
	case "social":
		score += 0.07
	case "security":
		score += 0.12
	default:
		score += 0.03
	}
	return round(minFloat(score, 0.98))
}

func impactBand(threatLevel int) string {
	if threatLevel >= 65 {
		return "High"
	}
	return "Moderate"
}

func sliceOfMaps(raw any) []map[string]any {
	items, ok := raw.([]any)
	if !ok {
		if typed, ok := raw.([]map[string]any); ok {
			return typed
		}
		return []map[string]any{}
	}
	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		if mapped, ok := item.(map[string]any); ok {
			result = append(result, mapped)
		}
	}
	return result
}

func stringValue(raw any) string {
	if value, ok := raw.(string); ok {
		return value
	}
	return ""
}

func stringOrDefault(raw any, fallback string) string {
	value := stringValue(raw)
	if value == "" {
		return fallback
	}
	return value
}

func number(raw any) float64 {
	switch value := raw.(type) {
	case float64:
		return value
	case float32:
		return float64(value)
	case int:
		return float64(value)
	case int64:
		return float64(value)
	case json.Number:
		parsed, _ := value.Float64()
		return parsed
	default:
		return 0
	}
}

func numberOrDefault(raw any, fallback float64) float64 {
	value := number(raw)
	if value == 0 {
		return fallback
	}
	return value
}

func round(value float64) float64 {
	return float64(int(value*100+0.5)) / 100
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func minFloat(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func maxFloat(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

File: apps/api/src/api_service/ai.py
"""AI gateway for the API service."""

from __future__ import annotations

import asyncio
import json
import logging
import os
import time
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any
from urllib import error, parse, request


class AIGatewayError(RuntimeError):
    pass


logger = logging.getLogger("api_service.ai_gateway")


@dataclass
class AIGateway:
    base_url: str = ""
    timeout_seconds: float = 5.0
    fallback_to_mock: bool = True

    def __post_init__(self) -> None:
        configured_url = self.base_url or os.getenv("AI_INFERENCE_URL", "http://localhost:8100").strip()
        self.base_url = configured_url.rstrip("/") if configured_url else ""
        self.fallback_to_mock = os.getenv("AI_FALLBACK_TO_MOCK", "true").strip().lower() not in {"0", "false", "no"}

    @property
    def remote_enabled(self) -> bool:
        return bool(self.base_url)

    @property
    def engine_name(self) -> str:
        if self.remote_enabled:
            return "contract_ai_service"
        return "mock_ai_service" if self.fallback_to_mock else "unconfigured_ai_service"

    async def process_signals(self, signals: list[dict[str, Any]]) -> dict[str, Any]:
        logger.info("process_signals called", extra={"signal_count": len(signals), "remote_enabled": self.remote_enabled})
        if not self.remote_enabled:
            if not self.fallback_to_mock:
                self._require_remote()
            logger.info("process_signals using mock fallback (remote disabled)")
            return self._mock_process_signals(signals)
        try:
            started_at = time.perf_counter()
            region_id = self._region_from_signals(signals)
            risk = await self._post_json(
                "/infer/risk",
                {"region_id": region_id, "at_time": datetime.utcnow().isoformat(), "signals": signals},
            )
            explain = await self._post_json(
                "/infer/explain",
                {"region_id": region_id, "risk_score": risk["risk_score"]},
            )
            interventions = await self._post_json(
                "/infer/interventions",
                {"region_id": region_id, "risk_score": risk["risk_score"]},
            )
            return {
                "assessment": {
                    "threat_level": int(round(float(risk["risk_score"]) * 100)),
                    "confidence": float(risk["confidence"]),
                    "risk_factors": [driver["name"] for driver in risk["top_drivers"]],
                },
                "indicators": [
                    {
                        "type": str(driver["name"]).lower().replace(" ", "_"),
                        "severity": round(float(driver["contribution"]), 4),
                        "description": f"{driver['name']} is driving current risk in {risk['region_id']}.",
                    }
                    for driver in risk["top_drivers"]
                ],
                "recommendations": [
                    {
                        "action": item["category"],
                        "priority": index + 1,
                        "estimated_impact": item["expected_impact"],
                    }
                    for index, item in enumerate(interventions["interventions"])
                ],
                "metadata": {
                    "model_version": risk["model_version"],
                    "processing_time": round(time.perf_counter() - started_at, 4),
                    "engine": self.engine_name,
                    "feature_snapshot_timestamp": risk["feature_snapshot_timestamp"],
                    "known_data_gaps": risk["known_data_gaps"],
                    "explanation_summary": explain["summary"],
                },
            }
        except AIGatewayError:
            if not self.fallback_to_mock:
                raise
            logger.warning("process_signals remote call failed; using mock fallback")
            return self._mock_process_signals(signals)

    async def predict_risk(self, region: str, timeframe: str) -> dict[str, Any]:
        logger.info("predict_risk called", extra={"region": region, "timeframe": timeframe, "remote_enabled": self.remote_enabled})
        if not self.remote_enabled:
            if not self.fallback_to_mock:
                self._require_remote()
            logger.info("predict_risk using mock fallback (remote disabled)", extra={"region": region})
            return self._mock_predict_risk(region, timeframe)
        try:
            risk = await self._post_json(
                "/infer/risk",
                {"region_id": region, "at_time": datetime.utcnow().isoformat(), "signals": []},
            )
            days = self._parse_days(timeframe)
            predictions = [
                {
                    "date": (datetime.utcnow().date() + timedelta(days=offset)).isoformat(),
                    "risk_level": int(round(float(risk["risk_score"]) * 100)),
                    "confidence": round(float(risk["confidence"]), 2),
                }
                for offset in range(min(days, 7))
            ]
            return {
                "region": risk["region_id"],
                "timeframe": timeframe,
                "predictions": predictions,
                "confidence": risk["confidence"],
                "factors": [driver["name"] for driver in risk["top_drivers"]],
                "known_data_gaps": risk["known_data_gaps"],
                "metadata": {
                    "model_version": risk["model_version"],
                    "engine": self.engine_name,
                    "feature_snapshot_timestamp": risk["feature_snapshot_timestamp"],
                },
            }
        except AIGatewayError:
            if not self.fallback_to_mock:
                raise
            logger.warning("predict_risk remote call failed; using mock fallback", extra={"region": region})
            return self._mock_predict_risk(region, timeframe)

    async def analyze_drivers(self, payload: dict[str, Any]) -> dict[str, Any]:
        logger.info("analyze_drivers called", extra={"region": payload.get("region_id", "national"), "remote_enabled": self.remote_enabled})
        if not self.remote_enabled:
            if not self.fallback_to_mock:
                self._require_remote()
            logger.info("analyze_drivers using mock fallback (remote disabled)")
            return self._mock_driver_analysis(payload)
        try:
            explain = await self._post_json(
                "/infer/explain",
                {"region_id": payload.get("region_id", "national"), "risk_score": payload.get("risk_score")},
            )
            drivers = [
                {
                    "name": driver["name"],
                    "contribution": round(float(driver["contribution"]), 4),
                    "trend": "increasing" if driver["direction"] == "up" else "decreasing",
                    "confidence": 0.8,
                }
                for driver in explain["top_drivers"]
            ]
            relationships = []
            if len(drivers) >= 2:
                relationships.append(
                    {
                        "from": drivers[0]["name"],
                        "to": drivers[1]["name"],
                        "strength": round(max(drivers[0]["contribution"], drivers[1]["contribution"]), 4),
                    }
                )
            return {
                "drivers": drivers,
                "causal_relationships": relationships,
                "metadata": {
                    "model_version": explain["model_version"],
                    "engine": self.engine_name,
                    "uncertainty_notes": explain["uncertainty_notes"],
                    "feature_snapshot_timestamp": explain["feature_snapshot_timestamp"],
                },
            }
        except AIGatewayError:
            if not self.fallback_to_mock:
                raise
            logger.warning("analyze_drivers remote call failed; using mock fallback")
            return self._mock_driver_analysis(payload)

    async def recommend_interventions(self, region: str, risk_profile: dict[str, Any]) -> dict[str, Any]:
        logger.info("recommend_interventions called", extra={"region": region, "remote_enabled": self.remote_enabled})
        if not self.remote_enabled:
            if not self.fallback_to_mock:
                self._require_remote()
            logger.info("recommend_interventions using mock fallback (remote disabled)", extra={"region": region})
            return self._mock_interventions(region, risk_profile)
        try:
            risk_score = float(risk_profile.get("risk_score", risk_profile.get("threat_level", 0) / 100))
            response = await self._post_json(
                "/infer/interventions",
                {"region_id": region, "risk_score": risk_score},
            )
            return {
                "region": response["region_id"],
                "interventions": [
                    {
                        "category": item["category"],
                        "expectedImpact": item["expected_impact"],
                        "timeToEffect": item["time_to_effect"],
                        "costBand": "Policy-defined",
                        "confidence": item["confidence"],
                        "constraintsApplied": item["constraints_applied"],
                    }
                    for item in response["interventions"]
                ],
                "metadata": {
                    "model_version": response["model_version"],
                    "engine": self.engine_name,
                    "feature_snapshot_timestamp": response["feature_snapshot_timestamp"],
                },
            }
        except AIGatewayError:
            if not self.fallback_to_mock:
                raise
            logger.warning("recommend_interventions remote call failed; using mock fallback", extra={"region": region})
            return self._mock_interventions(region, risk_profile)

    async def list_regions(self, limit: int = 50) -> dict[str, Any]:
        logger.info("list_regions called", extra={"limit": limit, "remote_enabled": self.remote_enabled})
        if not self.remote_enabled:
            if not self.fallback_to_mock:
                self._require_remote()
            logger.info("list_regions using mock fallback (remote disabled)")
            return self._mock_regions(limit)
        try:
            return await self._get_json("/infer/regions", {"limit": limit})
        except AIGatewayError:
            if not self.fallback_to_mock:
                raise
            logger.warning("list_regions remote call failed; using mock fallback")
            return self._mock_regions(limit)

    async def _post_json(self, path: str, payload: dict[str, Any]) -> dict[str, Any]:
        self._require_remote()

        def send() -> dict[str, Any]:
            body = json.dumps(payload).encode("utf-8")
            req = request.Request(
                f"{self.base_url}{path}",
                data=body,
                headers={"Content-Type": "application/json"},
                method="POST",
            )
            try:
                with request.urlopen(req, timeout=self.timeout_seconds) as response:
                    return json.loads(response.read().decode("utf-8"))
            except (error.URLError, error.HTTPError, TimeoutError, json.JSONDecodeError) as exc:
                raise AIGatewayError(str(exc)) from exc

        return await asyncio.to_thread(send)

    async def _get_json(self, path: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        self._require_remote()

        def send() -> dict[str, Any]:
            query = f"?{parse.urlencode(params)}" if params else ""
            req = request.Request(f"{self.base_url}{path}{query}", method="GET")
            try:
                with request.urlopen(req, timeout=self.timeout_seconds) as response:
                    return json.loads(response.read().decode("utf-8"))
            except (error.URLError, error.HTTPError, TimeoutError, json.JSONDecodeError) as exc:
                raise AIGatewayError(str(exc)) from exc

        return await asyncio.to_thread(send)

    def _require_remote(self) -> None:
        if not self.remote_enabled:
            raise AIGatewayError("AI inference URL is not configured")

    def _region_from_signals(self, signals: list[dict[str, Any]]) -> str:
        for signal in signals:
            location = signal.get("location") or {}
            region = location.get("region") or location.get("name")
            if region:
                return str(region)
        return "national"

    def _parse_days(self, timeframe: str) -> int:
        cleaned = (timeframe or "7d").strip().lower()
        if cleaned.endswith("d") and cleaned[:-1].isdigit():
            return max(1, int(cleaned[:-1]))
        return 7

    def _mock_regions(self, limit: int) -> dict[str, Any]:
        timestamp = datetime.utcnow().isoformat()
        rows = [
            {
                "region_id": "north-frontier",
                "risk_score": 0.72,
                "confidence": 0.82,
                "confidence_band": "high",
                "primary_driver": "Resource stress",
                "feature_snapshot_timestamp": timestamp,
            },
            {
                "region_id": "coastal-belt",
                "risk_score": 0.58,
                "confidence": 0.79,
                "confidence_band": "medium",
                "primary_driver": "Market volatility",
                "feature_snapshot_timestamp": timestamp,
            },
            {
                "region_id": "capital-corridor",
                "risk_score": 0.44,
                "confidence": 0.77,
                "confidence_band": "medium",
                "primary_driver": "Service disruption",
                "feature_snapshot_timestamp": timestamp,
            },
            {
                "region_id": "lake-region",
                "risk_score": 0.39,
                "confidence": 0.75,
                "confidence_band": "medium",
                "primary_driver": "Mobility pressure",
                "feature_snapshot_timestamp": timestamp,
            },
        ]
        return {
            "regions": rows[: max(limit, 1)],
            "model_version": "mock-model-v1",
            "feature_count": 6,
            "training_rows": 60,
        }

    def _mock_predict_risk(self, region: str, timeframe: str) -> dict[str, Any]:
        target_region = region or "national"
        today = datetime.utcnow().date()
        days = min(self._parse_days(timeframe), 7)
        base = 68 if target_region in {"national", "north-frontier"} else 56
        predictions = [
            {
                "date": (today + timedelta(days=offset)).isoformat(),
                "risk_level": min(95, base + offset),
                "confidence": 0.8,
            }
            for offset in range(days)
        ]
        return {
            "region": target_region,
            "timeframe": timeframe,
            "predictions": predictions,
            "confidence": 0.8,
            "factors": ["Resource stress", "Market volatility", "Service disruption"],
            "known_data_gaps": [],
            "metadata": {
                "model_version": "mock-model-v1",
                "engine": self.engine_name,
                "feature_snapshot_timestamp": datetime.utcnow().isoformat(),
            },
        }

    def _mock_driver_analysis(self, payload: dict[str, Any]) -> dict[str, Any]:
        region = str(payload.get("region_id", "national"))
        return {
            "drivers": [
                {"name": "Resource stress", "contribution": 0.34, "trend": "increasing", "confidence": 0.84},
                {"name": "Market volatility", "contribution": 0.27, "trend": "increasing", "confidence": 0.81},
                {"name": "Service disruption", "contribution": 0.19, "trend": "decreasing", "confidence": 0.78},
            ],
            "causal_relationships": [
                {"from": "Resource stress", "to": "Market volatility", "strength": 0.34}
            ],
            "metadata": {
                "model_version": "mock-model-v1",
                "engine": self.engine_name,
                "uncertainty_notes": [f"Mock driver profile for {region}."],
                "feature_snapshot_timestamp": datetime.utcnow().isoformat(),
            },
        }

    def _mock_interventions(self, region: str, risk_profile: dict[str, Any]) -> dict[str, Any]:
        risk_score = float(risk_profile.get("risk_score", risk_profile.get("threat_level", 0) / 100))
        confidence = "High" if risk_score >= 0.65 else "Medium"
        return {
            "region": region or "national",
            "interventions": [
                {
                    "category": "Economic stabilization support",
                    "expectedImpact": "High",
                    "timeToEffect": "2-6 weeks",
                    "costBand": "Policy-defined",
                    "confidence": confidence,
                    "constraintsApplied": ["budget_ceiling", "equity_screen"],
                },
                {
                    "category": "Targeted community mediation",
                    "expectedImpact": "Medium",
                    "timeToEffect": "1-3 weeks",
                    "costBand": "Policy-defined",
                    "confidence": "Medium",
                    "constraintsApplied": ["civilian_protection_rule"],
                },
            ],
            "metadata": {
                "model_version": "mock-model-v1",
                "engine": self.engine_name,
                "feature_snapshot_timestamp": datetime.utcnow().isoformat(),
            },
        }

    def _mock_process_signals(self, signals: list[dict[str, Any]]) -> dict[str, Any]:
        region_id = self._region_from_signals(signals)
        prediction = self._mock_predict_risk(region_id, "7d")
        first = prediction["predictions"][0]
        interventions = self._mock_interventions(region_id, {"risk_score": first["risk_level"] / 100})
        drivers = self._mock_driver_analysis({"region_id": region_id})["drivers"]
        return {
            "assessment": {
                "threat_level": first["risk_level"],
                "confidence": prediction["confidence"],
                "risk_factors": [row["name"] for row in drivers],
            },
            "indicators": [
                {
                    "type": row["name"].lower().replace(" ", "_"),
                    "severity": row["contribution"],
                    "description": f"{row['name']} is elevating near-term risk in {region_id}.",
                }
                for row in drivers
            ],
            "recommendations": [
                {
                    "action": row["category"],
                    "priority": idx + 1,
                    "estimated_impact": row["expectedImpact"],
                }
                for idx, row in enumerate(interventions["interventions"])
            ],
            "metadata": {
                "model_version": "mock-model-v1",
                "processing_time": 0.01,
                "engine": self.engine_name,
                "feature_snapshot_timestamp": datetime.utcnow().isoformat(),
                "known_data_gaps": [],
                "explanation_summary": "Mock response used because remote AI service is unavailable.",
            },
        }
